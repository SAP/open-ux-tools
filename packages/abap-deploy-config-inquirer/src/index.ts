import { ToolsLogger, type Logger } from '@sap-ux/logger';
import { initI18n } from './i18n';
import { PromptState } from './prompts/prompt-state';
import { getAbapDeployConfigQuestions } from './prompts';
import LoggerHelper from './logger-helper';
import type { InquirerAdapter } from '@sap-ux/inquirer-common';
import type { AbapDeployConfigAnswers, AbapDeployConfigPromptOptions } from './types';

/**
 * Get the inquirer prompts for abap deploy config.
 *
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompts used to provide input for abap deploy config generation and a reference to the answers object which will be populated with the user's responses once `inquirer.prompt` returns
 */
async function getPrompts(
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<{ prompts: any[]; answers: Partial<AbapDeployConfigAnswers> }> {
    await initI18n();
    LoggerHelper.logger = logger ?? new ToolsLogger({ logPrefix: '@sap-ux/abap-deploy-config-inquirer' });

    return {
        prompts: await getAbapDeployConfigQuestions(promptOptions),
        // Return reference to derived answers object that will be populated with user responses (after prompting is complete)
        answers: PromptState.abapDeployConfig
    };
}

/**
 * Prompt for abap deploy config writer inputs.
 *
 * @param adapter - optionally provide references to a calling inquirer instance, this supports integration to Yeoman generators, for example
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompt answers
 */
async function prompt(
    adapter: InquirerAdapter,
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<AbapDeployConfigAnswers> {
    const abapDeployConfigPrompts = (await getPrompts(promptOptions, logger)).prompts;
    const answers = await adapter.prompt<AbapDeployConfigAnswers>(abapDeployConfigPrompts);

    // Add dervied service answers to the answers object
    Object.assign(answers, PromptState.abapDeployConfig);

    return answers;
}

export { getPrompts, prompt, type InquirerAdapter, type AbapDeployConfigAnswers, type AbapDeployConfigPromptOptions };
