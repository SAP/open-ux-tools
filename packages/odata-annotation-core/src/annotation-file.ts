import type {
    Element,
    Attribute,
    ElementName,
    AttributeName,
    TextNode,
    AnyNode
} from '@sap-ux/odata-annotation-core-types';
import { ELEMENT_TYPE, Edm, TEXT_TYPE } from '@sap-ux/odata-annotation-core-types';

export const elements = (predicate: (child: Element) => boolean | undefined, element: Element): Element[] =>
    (element.content || []).filter(
        (content) => content.type === ELEMENT_TYPE && (!predicate || predicate(content))
    ) as Element[];

export const elementsWithName = (name: ElementName, element: Element): Element[] =>
    elements((content: Element) => content.name === name, element);

export const getElementAttribute = (element: Element, name: AttributeName): Attribute | undefined =>
    element?.attributes?.[name];

export const getElementAttributeValue = (element: Element, name: AttributeName): string => {
    const attributeNode = getElementAttribute(element, name);
    return attributeNode?.value ?? '';
};

/**
 * Determines whether the node is of element type with given name.
 *
 * @param node node to check
 * @param name name to check
 * @returns boolean check result
 */
export function isElementWithName(node: AnyNode | undefined, name: string): node is Element {
    return node?.type === ELEMENT_TYPE && node.name === name;
}

/**
 * Returns true if text doesn't contain any visible charactes (tabs, spaces, line breaks are ignored).
 *
 * @param text
 * @returns boolean check result
 */
function isEmptyText(text: string): boolean {
    return (text || '').replace(/[\s]/g, '').length === 0;
}

/**
 * Get text node content of an element.
 * Elements content is supposed to only contain 'Annotation' tags or single text node.
 *
 * @param element
 * @returns TextNode
 */
export function getSingleTextNode(element: Element): TextNode | null {
    let isInvalid = false;
    let firstTextNode: TextNode | null = null;
    (element.content || []).forEach((node) => {
        if (!isInvalid) {
            if (node.type === ELEMENT_TYPE && node.name !== Edm.Annotation) {
                isInvalid = true; // child element which is not 'Annotation'
            } else if (node.type === TEXT_TYPE && !isEmptyText(node.text)) {
                if (firstTextNode) {
                    isInvalid = true; // second text node
                } else {
                    firstTextNode = node;
                }
            }
        }
    });
    return isInvalid ? null : firstTextNode;
}
