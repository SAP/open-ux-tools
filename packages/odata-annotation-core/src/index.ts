export {
    ParsedName,
    ParsedActionFunctionSignature,
    ParsedCollectionIdentifier,
    ParsedIdentifier,
    parseIdentifier,
    toFullyQualifiedName,
    resolveName,
    getAliasInformation,
    getAllNamespacesAndReferences,
    toAliasQualifiedName
} from './names';

export {
    NavigationPropertyAnnotationSegment,
    ParsedPath,
    ParsedPathSegment,
    TermCastSegment,
    parsePath,
    toFullyQualifiedPath
} from './paths';

export { findPathToPosition, FindPathResult, getPositionData } from './search';

export * from './position';

export {
    elements,
    elementsWithName,
    getElementAttribute,
    getElementAttributeValue,
    getSingleTextNode,
    isElementWithName
} from './annotation-file';

export * from './utils';

export * from '@sap-ux/odata-annotation-core-types';
