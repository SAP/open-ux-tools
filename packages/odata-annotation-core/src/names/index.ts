export {
    ParsedName,
    ParsedActionFunctionSignature,
    ParsedCollectionIdentifier,
    ParsedIdentifier,
    parseIdentifier,
    COLLECTION_PREFIX
} from './parse';
export { toFullyQualifiedName, resolveName, toAliasQualifiedName } from './normalization';

export { getAliasInformation, getAllNamespacesAndReferences } from './namespaces';
