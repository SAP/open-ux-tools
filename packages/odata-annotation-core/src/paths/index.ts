export { toFullyQualifiedPath } from './normalization';
export {
    NavigationPropertyAnnotationSegment,
    ParsedPath,
    ParsedPathSegment,
    TermCastSegment,
    parsePath
} from './parse';
