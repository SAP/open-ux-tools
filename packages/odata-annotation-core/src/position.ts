import type { Position, Range } from '@sap-ux/odata-annotation-core-types';

/**
 * Checks if position1 is before position2.
 *
 * @param pos1 position1
 * @param pos2 position2
 * @param includeEqual when positions are equal - if true then result is true, otherwise false
 * @returns boolean result
 */
export function isBefore(pos1: Position, pos2: Position, includeEqual = false): boolean {
    if (pos1.line < pos2.line) {
        return true;
    }
    if (pos1.line > pos2.line) {
        return false;
    }
    if (includeEqual) {
        return pos1.character <= pos2.character;
    }

    return pos1.character < pos2.character;
}

/**
 * Checks if position is contained in range.
 *
 * @param range range
 * @param position position
 * @returns boolean result
 */
export function positionContained(range: Range | undefined, position: Position): range is Range {
    return range !== undefined && !isBefore(position, range.start, false) && isBefore(position, range.end, true);
}

/**
 * Checks if position is contained in range (range must be defined).
 *
 * @param range range
 * @param position position
 * @returns boolean result
 */
export function positionContainedStrict(range: Range, position: Position): boolean {
    return !isBefore(position, range.start, false) && isBefore(position, range.end, true);
}
/**
 * Check if the second range is within the first.
 *
 * @param a first range
 * @param b second range
 * @returns booelan result
 */
export function rangeContained(a: Range, b: Range): boolean {
    return isBefore(a.start, b.start, true) && isBefore(b.end, a.end, true);
}

/**
 * Get indent level based on the start position and tab width.
 *
 * @param startPosition
 * @param tabWidth
 * @returns numeric indent level
 */
export function getIndentLevel(startPosition: number, tabWidth: number): number {
    let level: number;
    if (startPosition < 0) {
        level = -1;
    } else {
        level = startPosition / tabWidth;
    }
    return level;
}

/**
 * Indents based on tabs or tab width.
 *
 * @param tabWidth
 * @param useTabs
 * @param level
 * @returns intentation string
 */
export function indent(tabWidth: number, useTabs: boolean, level: number): string {
    if (useTabs) {
        return '\t'.repeat(level);
    } else {
        return ' '.repeat(tabWidth * level);
    }
}
