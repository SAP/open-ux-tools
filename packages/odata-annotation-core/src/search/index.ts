export * from './find-by-position';
