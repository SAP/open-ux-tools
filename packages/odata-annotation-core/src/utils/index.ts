export * from './metadata';
export * from './metadata-service';
