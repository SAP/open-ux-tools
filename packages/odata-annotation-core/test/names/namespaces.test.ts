import type { Namespace, Reference } from '@sap-ux/odata-annotation-core-types';
import { getAliasInformation, getAllNamespacesAndReferences } from '../../src';

describe('names/namespaces.ts', () => {
    test('getAliasInformation', () => {
        const namespace: Namespace = {
            type: 'namespace',
            name: 'test'
        };
        const references: Reference[] = [
            {
                type: 'reference',
                name: 'com.sap.vocabularies.Common.v1',
                alias: 'Common'
            },
            {
                type: 'reference',
                name: 'com.sap.vocabularies.UI.v1',
                alias: 'UI'
            },
            {
                type: 'reference',
                name: 'com.sap.vocabularies.Communication.v1',
                alias: 'Communication'
            },
            {
                type: 'reference',
                name: 'SEPMRA_PROD_MAN',
                alias: 'SEPMRA_PROD_MAN'
            }
        ];
        const metadataNamespaces = new Set<string>().add('test');
        const result = getAliasInformation([namespace, ...references], metadataNamespaces);
        expect(result).toMatchSnapshot();
    });
});

describe('names/namespaces.ts', () => {
    test('getAllNamespacesAndReferences', () => {
        const namespace: Namespace = {
            type: 'namespace',
            name: 'test'
        };
        const references: Reference[] = [
            {
                type: 'reference',
                name: 'com.sap.vocabularies.Common.v1',
                alias: 'Common'
            },
            {
                type: 'reference',
                name: 'com.sap.vocabularies.UI.v1',
                alias: 'UI'
            },
            {
                type: 'reference',
                name: 'com.sap.vocabularies.Communication.v1',
                alias: 'Communication'
            },
            {
                type: 'reference',
                name: 'SEPMRA_PROD_MAN',
                alias: 'SEPMRA_PROD_MAN'
            }
        ];
        const result = getAllNamespacesAndReferences(namespace, references);
        expect(result).toMatchSnapshot();
    });
});
