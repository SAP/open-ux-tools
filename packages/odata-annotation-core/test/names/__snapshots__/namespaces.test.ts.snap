// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`names/namespaces.ts getAliasInformation 1`] = `
Object {
  "aliasMap": Object {
    "Common": "com.sap.vocabularies.Common.v1",
    "Communication": "com.sap.vocabularies.Communication.v1",
    "SEPMRA_PROD_MAN": "SEPMRA_PROD_MAN",
    "UI": "com.sap.vocabularies.UI.v1",
    "com.sap.vocabularies.Common.v1": "com.sap.vocabularies.Common.v1",
    "com.sap.vocabularies.Communication.v1": "com.sap.vocabularies.Communication.v1",
    "com.sap.vocabularies.UI.v1": "com.sap.vocabularies.UI.v1",
    "test": "test",
  },
  "aliasMapMetadata": Object {
    "test": "test",
  },
  "aliasMapVocabulary": Object {
    "Common": "com.sap.vocabularies.Common.v1",
    "Communication": "com.sap.vocabularies.Communication.v1",
    "SEPMRA_PROD_MAN": "SEPMRA_PROD_MAN",
    "UI": "com.sap.vocabularies.UI.v1",
    "com.sap.vocabularies.Common.v1": "com.sap.vocabularies.Common.v1",
    "com.sap.vocabularies.Communication.v1": "com.sap.vocabularies.Communication.v1",
    "com.sap.vocabularies.UI.v1": "com.sap.vocabularies.UI.v1",
  },
  "currentFileAlias": "",
  "currentFileNamespace": "test",
  "reverseAliasMap": Object {
    "SEPMRA_PROD_MAN": "SEPMRA_PROD_MAN",
    "com.sap.vocabularies.Common.v1": "Common",
    "com.sap.vocabularies.Communication.v1": "Communication",
    "com.sap.vocabularies.UI.v1": "UI",
    "test": "test",
  },
}
`;

exports[`names/namespaces.ts getAllNamespacesAndReferences 1`] = `
Array [
  Object {
    "name": "test",
    "type": "namespace",
  },
  Object {
    "alias": "Common",
    "name": "com.sap.vocabularies.Common.v1",
    "type": "reference",
  },
  Object {
    "alias": "UI",
    "name": "com.sap.vocabularies.UI.v1",
    "type": "reference",
  },
  Object {
    "alias": "Communication",
    "name": "com.sap.vocabularies.Communication.v1",
    "type": "reference",
  },
  Object {
    "alias": "SEPMRA_PROD_MAN",
    "name": "SEPMRA_PROD_MAN",
    "type": "reference",
  },
]
`;
